Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -27,130 +27,99 @@ private lemma cube_of_castHom_ne_zero {n : ZMod 9} :
   rcases cube_of_not_dvd hdvd.1.1 with ha | ha <;>
   rcases cube_of_not_dvd hdvd.1.2 with hb | hb <;>
   rcases cube_of_not_dvd hdvd.2 with hc | hc <;>
   rw [ha, hb, hc] <;> decide
 
-variable {a b c : ℕ} (hF : a ^ 3 + b ^ 3 = c ^ 3)
+end case1
 
-lemma three_dvd_gcd_of_dvd_a_of_dvd_c (ha : 3 ∣ a) (hc : 3 ∣ c) :
-    3 ∣ ({a, b, c} : Finset ℕ).gcd id := by
-  have hb : 3 ∣ b := by
-    have : 3 ∣ (b : ℤ) ^ 3 := by
-      replace hF : (a : ℤ) ^ 3 + (b : ℤ) ^ 3 = (c : ℤ) ^ 3 := by exact_mod_cast hF
-      rw [add_comm, ← eq_sub_iff_add_eq] at hF
-      rw [hF]
-      exact dvd_sub (dvd_pow (by exact_mod_cast hc) (by decide))
-        (dvd_pow (by exact_mod_cast ha) (by decide))
-    exact Int.coe_nat_dvd.1 <| Int.prime_three.dvd_of_dvd_pow this
-  refine Finset.dvd_gcd (fun x hx ↦ ?_)
-  simp only [Finset.mem_insert, Finset.mem_singleton] at hx
-  rcases hx with (hx | hx | hx)
-  · exact hx ▸ ha
-  · exact hx ▸ hb
-  · exact hx ▸ hc
+section case2
 
-lemma three_dvd_gcd_of_dvd_a_of_dvd_b (ha : 3 ∣ a) (hb : 3 ∣ b) :
-    3 ∣ ({a, b, c} : Finset ℕ).gcd id := by
-  have hc : 3 ∣ c := by
-    have : 3 ∣ (c : ℤ) ^ 3 := by
-      replace hF : (a : ℤ) ^ 3 + (b : ℤ) ^ 3 = (c : ℤ) ^ 3 := by exact_mod_cast hF
-      rw [← hF]
-      exact dvd_add (dvd_pow (by exact_mod_cast ha) (by decide))
-        (dvd_pow (by exact_mod_cast hb) (by decide))
-    exact Int.coe_nat_dvd.1 <| Int.prime_three.dvd_of_dvd_pow this
-  refine Finset.dvd_gcd (fun x hx ↦ ?_)
-  simp only [Finset.mem_insert, Finset.mem_singleton] at hx
-  rcases hx with (hx | hx | hx)
-  · exact hx ▸ ha
-  · exact hx ▸ hb
-  · exact hx ▸ hc
+private lemma three_dvd_c_of_dvd_a_of_dvd_b {a b c : ℤ} (HF : a ^ 3 + b ^ 3 + c ^ 3 = 0)
+    (h3a : 3 ∣ a) (h3b : 3 ∣ b) : 3 ∣ c := by
+  refine Int.prime_three.dvd_of_dvd_pow (n := 3) (dvd_neg.1 ?_)
+  rw [add_eq_zero_iff_eq_neg] at HF
+  exact HF.symm ▸ dvd_add (dvd_pow h3a (by decide)) (dvd_pow h3b (by decide))
 
-lemma three_dvd_gcd_of_dvd_b_of_dvd_c (hb : 3 ∣ b) (hc : 3 ∣ c) :
-    3 ∣ ({a, b, c} : Finset ℕ).gcd id := by
-  have ha : 3 ∣ a := by
-    have : 3 ∣ (a : ℤ) ^ 3 := by
-      replace hF : (a : ℤ) ^ 3 + (b : ℤ) ^ 3 = (c : ℤ) ^ 3 := by exact_mod_cast hF
-      rw [← eq_sub_iff_add_eq] at hF
-      rw [hF]
-      exact dvd_sub (dvd_pow (by exact_mod_cast hc) (by decide))
-            (dvd_pow (by exact_mod_cast hb) (by decide))
-    exact Int.coe_nat_dvd.1 <| Int.prime_three.dvd_of_dvd_pow this
-  refine Finset.dvd_gcd (fun x hx ↦ ?_)
-  simp only [Finset.mem_insert, Finset.mem_singleton] at hx
-  rcases hx with (hx | hx | hx)
-  · exact hx ▸ ha
-  · exact hx ▸ hb
-  · exact hx ▸ hc
+private lemma three_dvd_b_of_dvd_a_of_gcd_eq_one_of_case2 {a b c : ℤ} (ha : a ≠ 0) (hc : c ≠ 0)
+    (Hgcd: Finset.gcd {a, b, c} id = 1) (h3a : 3 ∣ a) (HF : a ^ 3 + b ^ 3 + c ^ 3 = 0)
+    (H : ∀ a b c : ℤ, c ≠ 0 → ¬ 3 ∣ a → ¬ 3 ∣ b  → 3 ∣ c → IsCoprime a b → a ^ 3 + b ^ 3 ≠ c ^ 3) :
+    3 ∣ b := by
+  have hbc : IsCoprime (-b) (-c) := by
```

Comment:
Would it make sense to have a general lemma
```lean
lemma isCoprime_of_gcd_eq_one_of_FLT {n : ℕ} {a b c : ℤ} (Hgcd: Finset.gcd {a, b, c} id = 1)
    (HF : a ^ n + b ^ n + c ^ n = 0) :
    IsCoprime a b := ...
```
? I can imagine that this is likely to be used a number of times in the FLT project...

Reply:
I've added it (it needs that one of the variable is different from `0`, I picked `b`).

Reply:
Wait, maybe it doesn't need the assumption, let me see.

Reply:
It doesn't, not the prettiest proof. I will have a look tomorrow or later today.

Reply:
The proof is now better. If find amusing that we are missing `a ^ n ∣ b ^ n ↔ a ∣ b`.

Reply:
what about [Nat.pow_dvd_pow_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/GCD/Basic.html#Nat.pow_dvd_pow_iff) and [Int.pow_dvd_pow_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Int/GCD.html#Int.pow_dvd_pow_iff) ?

Reply:
Ops, I somehow didn't find it. Let me use it to reduce imports, but I will also keep the version for integrally closed domains.